react-redux/index.d.ts

```diff
@@ -6,102 +6,119 @@
 
-type ComponentClass<P> = React.ComponentClass<P>;
-type StatelessComponent<P> = React.StatelessComponent<P>;
-type ReactNode = React.ReactNode;
-type Store<S> = Redux.Store<S>;
-type Dispatch<S> = Redux.Dispatch<S>;
-type ActionCreator<A> = Redux.ActionCreator<A>;
-
-interface ComponentDecorator<TOriginalProps, TOwnProps> {
-    (component: ComponentClass<TOriginalProps> | StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
-}
+export = ReactRedux;
 
-/**
- * Decorator that infers the type from the original component
- *
- * Can't use the above decorator because it would default the type to {}
- */
-export interface InferableComponentDecorator {
-    <P, TComponentConstruct extends (ComponentClass<P> | StatelessComponent<P>)>(component: TComponentConstruct): TComponentConstruct;
-}
+declare namespace ReactRedux {
 
-/**
- * Connects a React component to a Redux store.
- *
- * - Without arguments, just wraps the component, without changing the behavior / props
- *
- * - If 2 params are passed (3rd param, mergeProps, is skipped), default behavior
- * is to override ownProps (as stated in the docs), so what remains is everything that's
- * not a state or dispatch prop
- *
- * - When 3rd param is passed, we don't know if ownProps propagate and whether they
- * should be valid component props, because it depends on mergeProps implementation.
- * As such, it is the user's responsibility to extend ownProps interface from state or
- * dispatch props or both when applicable
- *
- * @param mapStateToProps
- * @param mapDispatchToProps
- * @param mergeProps
- * @param options
- */
-export declare function connect(): InferableComponentDecorator;
-
-export declare function connect<TStateProps, TDispatchProps, TOwnProps>(
-    mapStateToProps: FuncOrSelf<MapStateToProps<TStateProps, TOwnProps>>,
-    mapDispatchToProps?: FuncOrSelf<MapDispatchToPropsFunction<TDispatchProps, TOwnProps> | MapDispatchToPropsObject>
-): ComponentDecorator<TStateProps & TDispatchProps, TOwnProps>;
-
-export declare function connect<TStateProps, TDispatchProps, TOwnProps>(
-    mapStateToProps: FuncOrSelf<MapStateToProps<TStateProps, TOwnProps>>,
-    mapDispatchToProps: FuncOrSelf<MapDispatchToPropsFunction<TDispatchProps, TOwnProps> | MapDispatchToPropsObject>,
-    mergeProps: MergeProps<TStateProps, TDispatchProps, TOwnProps>,
-    options?: Options
-): ComponentDecorator<TStateProps & TDispatchProps, TOwnProps>;
-
-type FuncOrSelf<T> = T | (() => T);
-
-interface MapStateToProps<TStateProps, TOwnProps> {
-    (state: any, ownProps?: TOwnProps): TStateProps;
-}
+    type ComponentType<P> = React.ComponentClass<P> | React.StatelessComponent<P>;
 
-interface MapDispatchToPropsFunction<TDispatchProps, TOwnProps> {
-    (dispatch: Dispatch<any>, ownProps?: TOwnProps): TDispatchProps;
-}
+    interface ComponentDecorator<M, P> {
+        <TargetClass extends ComponentType<M>>(component: TargetClass): TargetClass & ConnectClass<P, TargetClass>;
```

Comment:
I took a closer look at this, and I think my initial reading is correct.

If we take the second overload of `connect` and assume the user has specified `TStateProps`/`TDispatchProps`/`TOwnProps` explicitly

```ts
function connect<TStateProps, TDispatchProps, TOwnProps>(...): ComponentDecorator<DefaultMergedProps<TStateProps, TDispatchProps, TOwnProps>, TOwnProps>;
```

and trace the types through a bit

```ts
(component: ComponentType<DefaultMergedProps<TStateProps, TDispatchProps, TOwnProps>>): ComponentType<DefaultMergedProps<TStateProps, TDispatchProps, TOwnProps>> & ConnectClass<TOwnProps, ComponentType<DefaultMergedProps<TStateProps, TDispatchProps, TOwnProps>>>
```

the resulting component has the type

```ts
ComponentType<TOwnProps & TStateProps & TDispatchProps> & ConnectClass<TOwnProps, ComponentType<TOwnProps & TStateProps & TDispatchProps>>

// dropping the StatelessComponent variant for simplicity and removing extraneous bits of ConnectClass

React.ComponentClass<TOwnProps & TStateProps & TDispatchProps> & React.ComponentClass<TOwnProps>

// which Typescript will presumably simplify to something like the following, or close to it

React.ComponentClass<TOwnProps & TStateProps & TDispatchProps>
```

I think this is right, but please correct me if I'm wrong. I hope I didn't drop anything important in the simplification steps.